File: backend/src/utils/passwordPolicy.js
// src/utils/passwordPolicy.js
// Política de senhas. Ponto único: o mínimo, a lista de senhas proibidas e a
// regra dos perfis com poder vivem aqui, e não espalhados pelos schemas.
//
// Porquê: a plataforma guarda dinheiro de terceiros. Uma conta de COMPANY_ADMIN
// aprova ordens de compra e uma de FINANCEIRO autoriza pagamentos — 8 caracteres
// não chegam para isso. O mínimo geral também subiu: 8 é o que se pedia há dez
// anos, e hoje quebra-se por força bruta em minutos.
//
// A política aplica-se a quem DEFINE ou MUDA uma senha. Nunca ao login: quem já
// tem uma senha curta continua a entrar, e é convidado a trocá-la — bloquear o
// acesso a utilizadores existentes seria trocar um risco por uma avaria.

// Perfis que movem dinheiro ou credenciam empresas.
const PERFIS_SENSIVEIS = ['COMPANY_ADMIN', 'FINANCEIRO', 'ADMIN_SISTEMA'];

const MINIMO = 10;
const MINIMO_SENSIVEL = 12;

// As senhas mais usadas do mundo e as variantes óbvias com o nome da plataforma.
// Uma lista curta apanha a esmagadora maioria das escolhas más sem precisar de
// serviço externo nem de base de dados de fugas.
const PROIBIDAS = new Set([
  '12345678', '123456789', '1234567890', '123456789010', '12345678910',
  'password', 'password1', 'password123', 'passw0rd', 'senha123', 'senha1234',
  'qwerty123', 'qwertyuiop', 'abc123456', 'admin123', 'administrador',
  'iloveyou', 'welcome123', 'letmein123', 'sunshine1', 'football1',
  'kixima', 'kixima123', 'kixima2026', 'kixima@123', 'kiximakixima',
  'angola123', 'luanda123', 'petroleo123',
]);

function minimoPara(role) {
  return PERFIS_SENSIVEIS.includes(role) ? MINIMO_SENSIVEL : MINIMO;
}

// Sequências e repetições: "aaaaaaaaaa" ou "1234567890" têm o comprimento todo
// e nenhuma resistência.
function ePobre(senha) {
  const s = String(senha);
  if (/^(.)\1+$/.test(s)) return true;                       // um só carácter repetido
  const seq = '0123456789abcdefghijklmnopqrstuvwxyz';
  const min = s.toLowerCase();
  return seq.includes(min) || [...seq].reverse().join('').includes(min);
}

/**
 * Valida uma senha. Devolve null se serve, ou a razão pela qual não serve —
 * sempre a dizer o que fazer, não só o que está mal.
 * @param senha  a senha escolhida
 * @param role   perfil a que a conta se destina (opcional)
 * @param email  email da conta, para recusar a senha igual ao próprio email
 */
function validar(senha, { role, email } = {}) {
  const s = String(senha || '');
  const min = minimoPara(role);

  if (s.length < min) {
    return PERFIS_SENSIVEIS.includes(role)
      ? `Esta conta aprova operações com dinheiro, por isso a senha precisa de pelo menos ${min} caracteres.`
      : `A senha deve ter pelo menos ${min} caracteres.`;
  }
  if (PROIBIDAS.has(s.toLowerCase())) {
    return 'Esta senha é das mais usadas no mundo e é testada em primeiro lugar num ataque. Escolha outra.';
  }
  if (ePobre(s)) {
    return 'A senha não pode ser uma sequência nem um carácter repetido.';
  }
  if (email) {
    const local = String(email).split('@')[0].toLowerCase();
    if (local.length >= 4 && s.toLowerCase().includes(local)) {
      return 'A senha não pode conter o seu email.';
    }
  }
  return null;
}

module.exports = { validar, minimoPara, MINIMO, MINIMO_SENSIVEL, PERFIS_SENSIVEIS };

File: backend-java/src/main/java/ao/kixima/user/PasswordPolicy.java
package ao.kixima.user;

import ao.kixima.security.PersonaRole;
import org.springframework.stereotype.Component;

import java.util.Set;
import java.util.regex.Pattern;

/**
 * Espelha backend/src/utils/passwordPolicy.js — ponto único da política de
 * senhas. Aplica-se a quem DEFINE ou MUDA uma senha, nunca ao login (quem já
 * tem uma senha curta continua a entrar).
 */
@Component
public class PasswordPolicy {

    public static final Set<PersonaRole> PERFIS_SENSIVEIS = Set.of(
            PersonaRole.COMPANY_ADMIN, PersonaRole.FINANCEIRO, PersonaRole.ADMIN_SISTEMA);

    public static final int MINIMO = 10;
    public static final int MINIMO_SENSIVEL = 12;

    private static final Set<String> PROIBIDAS = Set.of(
            "12345678", "123456789", "1234567890", "123456789010", "12345678910",
            "password", "password1", "password123", "passw0rd", "senha123", "senha1234",
            "qwerty123", "qwertyuiop", "abc123456", "admin123", "administrador",
            "iloveyou", "welcome123", "letmein123", "sunshine1", "football1",
            "kixima", "kixima123", "kixima2026", "kixima@123", "kiximakixima",
            "angola123", "luanda123", "petroleo123");

    private static final Pattern UM_SO_CARACTER_REPETIDO = Pattern.compile("^(.)\\1+$");
    private static final String SEQUENCIA = "0123456789abcdefghijklmnopqrstuvwxyz";

    public int minimoPara(PersonaRole role) {
        return role != null && PERFIS_SENSIVEIS.contains(role) ? MINIMO_SENSIVEL : MINIMO;
    }

    private boolean ePobre(String senha) {
        String s = senha == null ? "" : senha;
        if (UM_SO_CARACTER_REPETIDO.matcher(s).matches()) return true;
        String min = s.toLowerCase();
        return SEQUENCIA.contains(min) || new StringBuilder(SEQUENCIA).reverse().toString().contains(min);
    }

    /**
     * Valida uma senha. Devolve null se serve, ou a razão pela qual não serve.
     *
     * @param senha a senha escolhida
     * @param role  perfil a que a conta se destina (pode ser null)
     * @param email email da conta, para recusar a senha igual ao próprio email (pode ser null)
     */
    public String validar(String senha, PersonaRole role, String email) {
        String s = senha == null ? "" : senha;
        int min = minimoPara(role);

        if (s.length() < min) {
            return role != null && PERFIS_SENSIVEIS.contains(role)
                    ? "Esta conta aprova operações com dinheiro, por isso a senha precisa de pelo menos " + min + " caracteres."
                    : "A senha deve ter pelo menos " + min + " caracteres.";
        }
        if (PROIBIDAS.contains(s.toLowerCase())) {
            return "Esta senha é das mais usadas no mundo e é testada em primeiro lugar num ataque. Escolha outra.";
        }
        if (ePobre(s)) {
            return "A senha não pode ser uma sequência nem um carácter repetido.";
        }
        if (email != null && !email.isBlank()) {
            String local = email.split("@")[0].toLowerCase();
            if (local.length() >= 4 && s.toLowerCase().contains(local)) {
                return "A senha não pode conter o seu email.";
            }
        }
        return null;
    }

    public String validar(String senha) {
        return validar(senha, null, null);
    }
}
